core/state/state_object.go

```diff
@@ -126,31 +131,32 @@ func (self *StateObject) Storage() Storage {
 }
 
 func (self *StateObject) GetState(key common.Hash) common.Hash {
-	strkey := key.Str()
-	value, exists := self.storage[strkey]
+	sv, exists := self.storage[key]
 	if !exists {
-		value = self.getAddr(key)
-		if (value != common.Hash{}) {
-			self.storage[strkey] = value
+		sv = storageValue{value: self.getAddr(key)}
+		if (sv.value != common.Hash{}) {
+			self.storage[key] = sv
 		}
 	}
 
-	return value
+	return sv.value
 }
 
 func (self *StateObject) SetState(k, value common.Hash) {
-	self.storage[k.Str()] = value
+	self.storage[k] = storageValue{value: value, dirty: true}
 	self.dirty = true
 }
 
 // Update updates the current cached storage to the trie
 func (self *StateObject) Update() {
-	for key, value := range self.storage {
-		if (value == common.Hash{}) {
-			self.trie.Delete([]byte(key))
-			continue
+	for key, sv := range self.storage {
+		if sv.dirty {
+			if (sv.value == common.Hash{}) {
+				self.trie.Delete(key[:])
+				continue
+			}
+			self.setAddr(key, sv.value)
```

Comment:
As discussed on gitter this will stay the way it is.Fixing this proper is going to require significant changes in the core which may lead to consensus failures. This shouldn't be changed for 1.4